src/unix/event.rs

```diff
@@ -0,0 +1,346 @@
+
+    /// Clone this EventConsumer.
+    pub fn try_clone(&self) -> result::Result<EventConsumer, io::Error> {
+        Ok(EventConsumer {
+            fd: self.fd.try_clone()?,
+        })
+    }
+}
+
+impl AsRawFd for EventConsumer {
+    fn as_raw_fd(&self) -> std::os::unix::prelude::RawFd {
+        self.fd.as_raw_fd()
+    }
+}
+
+impl FromRawFd for EventConsumer {
+    unsafe fn from_raw_fd(fd: std::os::unix::prelude::RawFd) -> Self {
+        EventConsumer {
+            fd: File::from_raw_fd(fd),
+        }
+    }
+}
+
+impl IntoRawFd for EventConsumer {
+    fn into_raw_fd(self) -> std::os::unix::prelude::RawFd {
+        self.fd.into_raw_fd()
+    }
+}
+
+#[cfg(not(any(target_os = "linux", target_os = "android")))]
+fn fcntl_getfl_and_setfl(fd: i32, flag: i32) -> std::result::Result<(), io::Error> {
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let flags = unsafe { libc::fcntl(fd, libc::F_GETFL) };
+    if flags < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let ret = unsafe { libc::fcntl(fd, libc::F_SETFL, flags | flag) };
+    if ret < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    Ok(())
+}
+
+fn fcntl_getfd_and_setfd(fd: i32, flag: i32) -> std::result::Result<(), io::Error> {
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let flags = unsafe { libc::fcntl(fd, libc::F_GETFD) };
+    if flags < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let ret = unsafe { libc::fcntl(fd, libc::F_SETFD, flags | flag) };
+    if ret < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    Ok(())
+}
+
+/// Create a new EventNotifier and EventConsumer using a pipe.
+///
+/// # Arguments
+///
+/// * `flag` - Flags to set on the file descriptor, such as `EventFlag::NONBLOCK` or `EventFlag::CLOEXEC`.
+///
+/// # Examples
+///
+/// ```
+/// use vmm_sys_util::event::{new_event_consumer_and_notifier, EventFlag};
+/// let (consumer, notifier) = new_event_consumer_and_notifier(EventFlag::NONBLOCK)
+///     .expect("Failed to create notifier and consumer");
+/// notifier.notify().unwrap();
+/// assert!(consumer.consume().is_ok());
+/// ```
+#[cfg(not(any(target_os = "linux", target_os = "android")))]
+pub fn new_event_consumer_and_notifier(
+    flag: EventFlag,
+) -> std::result::Result<(EventConsumer, EventNotifier), io::Error> {
+    // Use a pipe for non-Linux platforms.
+    use std::os::fd::RawFd;
+    let mut fds: [RawFd; 2] = [-1, -1];
+    // SAFETY: Safe because we check the fd is valid and check the return value.
+    let ret = unsafe { libc::pipe(fds.as_mut_ptr()) };
+    if ret < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    if flag.contains(EventFlag::NONBLOCK) {
+        fcntl_getfl_and_setfl(fds[0], libc::O_NONBLOCK)?;
```

Comment:
If this fail, who close the `fd`s we opened?

Reply:
Same for next block.

Reply:
Add these lines (https://github.com/rust-vmm/vmm-sys-util/pull/244/files#diff-b817124a7a414d687ff2826960adbb7bf304a2770b0f6ee70b3daa5ca4a16113R188) to close fd. In another PR, perhaps we can provide a pipe wrapper to better use Rust's RAII

Reply:
Or just create `EventConsumer` and `EventNotifier` after pipe() as I mentioned in https://github.com/rust-vmm/vmm-sys-util/pull/244#discussion_r2238823427